diff --git a/Analysis/include/Luau/Scope.h b/Analysis/include/Luau/Scope.h
index 90b22954..8e1c106e 100644
--- a/Analysis/include/Luau/Scope.h
+++ b/Analysis/include/Luau/Scope.h
@@ -52,7 +52,13 @@ struct Scope
     std::unordered_map<Name, Location> typeAliasLocations;
     std::unordered_map<Name, Location> typeAliasNameLocations;
     std::unordered_map<Name, ModuleName> importedModules; // Mapping from the name in the require statement to the internal moduleName.
-    std::unordered_map<Name, std::unordered_map<Name, TypeFun>> importedTypeBindings;
+    struct ImportedTypeNamespace
+    {
+        std::unordered_map<Name, TypeFun> members;
+        std::optional<TypeFun> fallback;
+    };
+
+    std::unordered_map<Name, ImportedTypeNamespace> importedTypeBindings;
 
     DenseHashSet2<Name> builtinTypeNames;
     void addBuiltinTypeBinding(const Name& name, const TypeFun& tyFun);
diff --git a/Analysis/src/AutocompleteCore.cpp b/Analysis/src/AutocompleteCore.cpp
index 81fcd360..a5a3ac94 100644
--- a/Analysis/src/AutocompleteCore.cpp
+++ b/Analysis/src/AutocompleteCore.cpp
@@ -798,7 +798,7 @@ AutocompleteEntryMap autocompleteModuleTypes(const Module& module, const ScopePt
     {
         if (auto it = scope->importedTypeBindings.find(std::string(moduleName)); it != scope->importedTypeBindings.end())
         {
-            for (const auto& [name, ty] : it->second)
+            for (const auto& [name, ty] : it->second.members)
                 result[name] = AutocompleteEntry{AutocompleteEntryKind::Type, ty.type};
 
             break;
diff --git a/Analysis/src/ConstraintGenerator.cpp b/Analysis/src/ConstraintGenerator.cpp
index c6540ab4..d967643a 100644
--- a/Analysis/src/ConstraintGenerator.cpp
+++ b/Analysis/src/ConstraintGenerator.cpp
@@ -1575,7 +1575,7 @@ ControlFlow ConstraintGenerator::visit(const ScopePtr& scope, AstStatLocal* stat
                 continue;
 
             const Name name{statLocal->vars.data[i]->name.value};
-            scope->importedTypeBindings[name] = module->exportedTypeBindings;
+            scope->importedTypeBindings[name].members = module->exportedTypeBindings;
             scope->importedModules[name] = moduleInfo->name;
 
             // Imported typeArguments of requires that transitively refer to current module have to be replaced with 'any'
@@ -1584,7 +1584,7 @@ ControlFlow ConstraintGenerator::visit(const ScopePtr& scope, AstStatLocal* stat
                 if (path.empty() || path.front() != moduleInfo->name)
                     continue;
 
-                for (auto& [name, tf] : scope->importedTypeBindings[name])
+                for (auto& [name, tf] : scope->importedTypeBindings[name].members)
                     tf = TypeFun{{}, {}, builtinTypes->anyType};
             }
         }
diff --git a/Analysis/src/FragmentAutocomplete.cpp b/Analysis/src/FragmentAutocomplete.cpp
index fda1fdd5..88c73de9 100644
--- a/Analysis/src/FragmentAutocomplete.cpp
+++ b/Analysis/src/FragmentAutocomplete.cpp
@@ -751,7 +751,7 @@ void cloneTypesFromFragment(
     {
         if (std::optional<TypeFun> tf = staleScope->lookupImportedType(mod, name))
         {
-            destScope->importedTypeBindings[mod].insert_or_assign(name, Luau::cloneIncremental(*tf, *destArena, cloneState, destScope));
+            destScope->importedTypeBindings[mod].members.insert_or_assign(name, Luau::cloneIncremental(*tf, *destArena, cloneState, destScope));
         }
     }
 
diff --git a/Analysis/src/Scope.cpp b/Analysis/src/Scope.cpp
index 86299516..f104062b 100644
--- a/Analysis/src/Scope.cpp
+++ b/Analysis/src/Scope.cpp
@@ -132,20 +132,17 @@ std::optional<TypeFun> Scope::lookupImportedType(const Name& moduleAlias, const
     while (scope)
     {
         auto it = scope->importedTypeBindings.find(moduleAlias);
-        if (it == scope->importedTypeBindings.end())
+        if (it != scope->importedTypeBindings.end())
         {
-            scope = scope->parent.get();
-            continue;
-        }
+            auto it2 = it->second.members.find(name);
+            if (it2 != it->second.members.end())
+                return it2->second;
 
-        auto it2 = it->second.find(name);
-        if (it2 == it->second.end())
-        {
-            scope = scope->parent.get();
-            continue;
+            if (it->second.fallback.has_value())
+                return it->second.fallback;
         }
 
-        return it2->second;
+        scope = scope->parent.get();
     }
 
     return std::nullopt;
diff --git a/Analysis/src/ToString.cpp b/Analysis/src/ToString.cpp
index 4cb95633..4d9780c5 100644
--- a/Analysis/src/ToString.cpp
+++ b/Analysis/src/ToString.cpp
@@ -144,7 +144,7 @@ static std::pair<bool, std::optional<Luau::Name>> canUseTypeNameInScope(ScopePtr
     {
         for (const auto& [importName, nameTable] : curr->importedTypeBindings)
         {
-            if (nameTable.count(name))
+            if (nameTable.members.count(name))
                 return {true, importName};
         }
 
diff --git a/Analysis/src/TypeInfer.cpp b/Analysis/src/TypeInfer.cpp
index 6f3498be..530a7690 100644
--- a/Analysis/src/TypeInfer.cpp
+++ b/Analysis/src/TypeInfer.cpp
@@ -1137,7 +1137,7 @@ ControlFlow TypeChecker::check(const ScopePtr& scope, const AstStatLocal& local)
 
                     if (ModulePtr module = resolver->getModule(moduleInfo->name))
                     {
-                        scope->importedTypeBindings[name] = module->exportedTypeBindings;
+                        scope->importedTypeBindings[name].members = module->exportedTypeBindings;
                         scope->importedModules[name] = moduleInfo->name;
 
                         // Imported typeArguments of requires that transitively refer to current module have to be replaced with 'any'
@@ -1145,7 +1145,7 @@ ControlFlow TypeChecker::check(const ScopePtr& scope, const AstStatLocal& local)
                         {
                             if (!path.empty() && path.front() == moduleInfo->name)
                             {
-                                for (auto& [name, tf] : scope->importedTypeBindings[name])
+                                for (auto& [name, tf] : scope->importedTypeBindings[name].members)
                                     tf = TypeFun{{}, {}, anyType};
                             }
                         }
